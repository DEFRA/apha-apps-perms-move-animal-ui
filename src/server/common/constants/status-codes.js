/**
 * @typedef {Record<string, number>} StatusCodes
 */
export const statusCodes = {
  ok: 200,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  imATeapot: 418,
  redirect: 302,
  serverError: 500,
  badGateway: 502,
  gatewayTimeout: 504
}
