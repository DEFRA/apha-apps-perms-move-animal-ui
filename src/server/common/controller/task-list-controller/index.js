import { getAuthOptions } from '../../helpers/auth/toggles-helper.js'
/** @import { ApplicationModel } from '../../model/application/application.js' */
/** @import { StateManager } from '../../model/state/state-manager.js' */

export class TaskListController {
  /** @type {string} */
  pageTitle

  /** @type {string} */
  heading

  /** @type {string} */
  buttonText

  /** @type {string} */
  path

  /** @type {string} */
  submitApplicationPath

  /**
   * @param {typeof StateManager} StateManager
   * @param {typeof ApplicationModel} ApplicationModel
   */
  constructor(StateManager, ApplicationModel) {
    this.StateManager = StateManager
    this.ApplicationModel = ApplicationModel
  }

  taskListGetHandler(req, h) {
    const visibleSections = Object.values(
      this.ApplicationModel.fromState(new this.StateManager(req).toState())
        .tasks
    )

    const gdsTasks = visibleSections.map((section) => {
      return buildGdsTaskItem(section.buildGdsTaskDetails(req))
    })

    const incompleteTasks =
      visibleSections.length -
      visibleSections.filter((section) => {
        return section.validate().isValid
      }).length

    return h.view('common/controller/task-list-controller/index', {
      pageTitle: this.pageTitle,
      heading: this.heading,
      gdsTasks,
      incompleteTasks,
      buttonText: this.buttonText
    })
  }

  /**
   * A GDS styled example task list page controller.
   * @satisfies {Partial<ServerRoute>}
   */
  taskListPostHandler(_req, h) {
    return h.redirect(this.submitApplicationPath)
  }

  taskListIncompleteGetHandler(_req, h) {
    const pageTitle =
      'You need to complete all of the sections before you review and submit'
    const heading = pageTitle
    const buttonText = 'Go to your application'
    const taskListUrl = this.path

    return h.view(
      'common/controller/task-list-controller/task-list-incomplete',
      {
        pageTitle,
        heading,
        buttonText,
        taskListUrl
      }
    )
  }

  plugin() {
    const path = this.path
    const taskListGetHandler = this.taskListGetHandler.bind(this)
    const taskListPostHandler = this.taskListPostHandler.bind(this)
    const taskListIncompleteGetHandler =
      this.taskListIncompleteGetHandler.bind(this)

    return {
      plugin: {
        name: `task-list-${this.constructor.name}`,

        register(server) {
          const options = {
            ...getAuthOptions(false)
          }

          server.route([
            {
              method: 'GET',
              path,
              handler: taskListGetHandler,
              options
            },
            {
              method: 'POST',
              path,
              handler: taskListPostHandler,
              options
            },

            {
              method: 'GET',
              path: `${path}/incomplete`,
              options,
              handler: taskListIncompleteGetHandler
            }
          ])
        }
      }
    }
  }
}

/**
 * @param {object} params - The parameters to configure the GDS task item.
 * @param {string} params.title - The title of the task item.
 * @param {string} params.initialLink - The URL link to navigate when the task is incomplete or not enabled.
 * @param {string} params.summaryLink - The URL link to navigate when the task is completed.
 * @param {boolean} params.isValid - Indicates whether the task is valid or completed.
 * @param {boolean} params.isEnabled - Indicates whether the task is enabled and can be interacted with.
 * @returns {object} A GDS task item object ready to be used in a nunjucks template.
 */
function buildGdsTaskItem({
  title,
  initialLink,
  summaryLink,
  isValid,
  isEnabled
}) {
  let status, href
  if (!isEnabled) {
    status = {
      text: 'Cannot start yet',
      classes: 'govuk-task-list__status--cannot-start-yet'
    }
  } else if (isValid) {
    status = {
      text: 'Completed',
      classes: ''
    }
    href = summaryLink
  } else {
    status = {
      tag: {
        text: 'Incomplete',
        classes: 'govuk-tag--blue'
      }
    }
    href = initialLink
  }

  return {
    title: {
      text: title
    },
    href,
    status
  }
}

/**
 * @import { ServerRoute } from '@hapi/hapi'
 */
