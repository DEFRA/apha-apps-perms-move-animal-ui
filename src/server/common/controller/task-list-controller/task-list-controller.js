import { getAuthOptions } from '../../helpers/auth/toggles-helper.js'

/** @import { StateManager } from '../../model/state/state-manager.js' */
/** @import { ApplicationModel } from '../../model/application/application.js' */
/** @import { ServerRoute } from '@hapi/hapi' */

export class TaskListController {
  /** @type {typeof ApplicationModel} */
  ApplicationModel

  /** @type {typeof StateManager} */
  StateManager

  /** @type {string} */
  pageTitleAndHeading

  /** @type {string} */
  buttonText

  /** @type {string} */
  urlPath

  /** @type {string} */
  submitUrlPath

  taskListGetHandler(req, h) {
    const applicationState = new this.StateManager(req).toState()
    const application = this.ApplicationModel.fromState(applicationState)
    const visibleSections = Object.values(application.tasks)

    const gdsTasks = visibleSections.map((section) => {
      return buildGdsTaskItem(section.buildGdsTaskDetails(applicationState))
    })

    const incompleteTasks =
      visibleSections.length -
      visibleSections.filter((section) => {
        return section.validate().isValid
      }).length

    return h.view('common/controller/task-list-controller/index.njk', {
      pageTitle: this.pageTitleAndHeading,
      heading: this.pageTitleAndHeading,
      gdsTasks,
      incompleteTasks,
      buttonText: this.buttonText
    })
  }

  taskListPostHandler(_req, h) {
    return h.redirect(this.submitUrlPath)
  }

  plugin() {
    const options = {
      ...getAuthOptions()
    }

    /** @type {ServerRoute[]} */
    const routes = [
      {
        method: 'GET',
        path: this.urlPath,
        handler: this.taskListGetHandler.bind(this),
        options
      },
      {
        method: 'POST',
        path: this.urlPath,
        handler: this.taskListPostHandler.bind(this),
        options
      }
    ]

    return {
      plugin: {
        name: `task-list-${this.constructor.name}`,
        register(server) {
          server.route(routes)
        }
      }
    }
  }
}

/**
 * @param {object} params - The parameters to configure the GDS task item.
 * @param {string} params.title - The title of the task item.
 * @param {string} params.initialLink - The URL link to navigate when the task is incomplete or not enabled.
 * @param {string} params.summaryLink - The URL link to navigate when the task is completed.
 * @param {boolean} params.isValid - Indicates whether the task is valid or completed.
 * @param {boolean} params.isEnabled - Indicates whether the task is enabled and can be interacted with.
 * @returns {object} A GDS task item object ready to be used in a nunjucks template.
 */
function buildGdsTaskItem({
  title,
  initialLink,
  summaryLink,
  isValid,
  isEnabled
}) {
  let status, href
  if (!isEnabled) {
    status = {
      text: 'Cannot start yet',
      classes: 'govuk-task-list__status--cannot-start-yet'
    }
  } else if (isValid) {
    status = {
      text: 'Completed',
      classes: ''
    }
    href = summaryLink
  } else {
    status = {
      tag: {
        text: 'Incomplete',
        classes: 'govuk-tag--blue'
      }
    }
    href = initialLink
  }

  return {
    title: {
      text: title
    },
    href,
    status
  }
}
