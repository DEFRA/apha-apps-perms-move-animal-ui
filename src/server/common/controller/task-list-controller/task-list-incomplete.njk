{% from "govuk/components/button/macro.njk" import govukButton %}
{% extends 'layouts/page.njk' %}

{% block content %}
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <h1 class="govuk-heading-l">{{ heading }}</h1>
        {{
          govukButton({
            text: buttonText,
            classes: "govuk-button--primary",
            href: taskListUrl
          })
        }}
      </div>
    </div>
  </div>
{% endblock %}
