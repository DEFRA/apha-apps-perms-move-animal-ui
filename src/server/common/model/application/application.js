import { DestinationSection } from '../section/destination/destination.js'
import { LicenceSection } from '../section/licence/licence.js'
import { OriginSection } from '../section/origin/origin.js'
import { validateApplication } from './validation.js'

/**
 * export @typedef {{
 * origin: OriginData | undefined;
 * licence: LicenceData | undefined;
 * destination: DestinationData | undefined;
 * }} ApplicationData
 * @import {OriginData} from '../section/origin/origin.js'
 * @import {LicenceData} from '../section/licence/licence.js'
 * @import {DestinationData} from '../section/destination/destination.js'
 */

export class ApplicationModel {
  _data

  constructor(data) {
    this._data = data
    Object.seal(this)
  }

  validate() {
    return validateApplication(this._data)
  }

  /**
   * Getter for tasks.
   * @returns {Array|Object} The data associated with each and every task in the application
   */
  get tasks() {
    return this._data
  }

  /**
   * @param {ApplicationData | undefined} state
   * @returns {ApplicationModel}
   */
  static fromState(state) {
    return new ApplicationModel({
      origin: OriginSection.fromState(state?.origin),
      licence: LicenceSection.fromState(state?.licence),
      destination: DestinationSection.fromState(state?.destination)
    })
  }
}
