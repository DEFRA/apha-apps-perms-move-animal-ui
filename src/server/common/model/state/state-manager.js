/** @import {Request} from "@hapi/hapi/lib/types/request.js" */

/**
 * @typedef {Record<string, any>} RawSectionState
 * @typedef {Record<string, RawSectionState>} RawApplicationState
 */

const sectionKeys = ['origin', 'destination', 'licence', 'biosecurity']

export class StateManager {
  /** @param {Request} request */
  constructor(request) {
    this._request = request
  }

  /** @returns {RawApplicationState} */
  toState() {
    return Object.fromEntries(
      sectionKeys.map((key) => [key, this._request.yar.get(key)])
    )
  }
}
