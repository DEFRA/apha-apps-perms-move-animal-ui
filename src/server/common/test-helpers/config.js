import { config } from '~/src/config/config.js'

/* global jest */

const configGet = config.get.bind(config)

/**
 * @param {string} key
 * @param {any} value
 */
export const spyOnConfig = (key, value) => {
  jest.spyOn(config, 'get').mockImplementation((name) => {
    if (name === key) {
      return value
    } else {
      return configGet(name)
    }
  })
  return config
}
