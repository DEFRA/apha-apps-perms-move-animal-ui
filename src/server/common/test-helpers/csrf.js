import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
