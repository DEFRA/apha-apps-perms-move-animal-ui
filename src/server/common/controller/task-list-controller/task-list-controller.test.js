import { createServer } from '~/src/server/index.js'
import SessionTester from '~/src/server/common/test-helpers/session-helper.js'
import { statusCodes } from '~/src/server/common/constants/status-codes.js'
import { withCsrfProtection } from '~/src/server/common/test-helpers/csrf.js'
import { TaskListController } from './task-list-controller.js'
import { ApplicationModel } from '../../model/application/application.js'
import { StateManager } from '../../model/state/state-manager.js'
import { OriginSection } from '~/src/server/tb/origin/section.js'
import { DestinationSection } from '~/src/server/tb/destination/section.js'
import {
  validDestinationSectionState,
  validOriginSectionState
} from '../../test-helpers/journey-state.js'
import { parseDocument } from '../../test-helpers/dom.js'
import { spyOnConfig } from '../../test-helpers/config.js'

/** @import { Server } from '@hapi/hapi' */

const stateKey = 'test-state'

class TestApplicationModel extends ApplicationModel {
  static implementedSections = [OriginSection, DestinationSection]
}

class TestStateManager extends StateManager {
  get key() {
    return stateKey
  }
}

class TestTaskListController extends TaskListController {
  ApplicationModel = TestApplicationModel
  StateManager = TestStateManager
  pageTitleAndHeading = 'Test page heading'
  buttonText = 'Review your application and submit'
  urlPath = '/test/task-list'
  submitUrlPath = '/test/submit'
}

const controller = new TestTaskListController()

describe('TaskListController', () => {
  /** @type {Server} */
  let server

  /** @type {SessionTester} */
  let session

  beforeAll(async () => {
    server = await createServer()
    await server.register(controller.plugin())
    await server.initialize()
  })

  beforeEach(async () => {
    session = await SessionTester.create(server)
  })

  afterEach(jest.restoreAllMocks)

  afterAll(async () => {
    await server.stop({ timeout: 0 })
  })

  it('should have a unique plugin name based on the class', () => {
    expect(controller.plugin().plugin.name).toBe(
      'task-list-TestTaskListController'
    )
  })

  it("should have no auth options if auth isn't required", () => {
    spyOnConfig('featureFlags', {
      authEnabled: false,
      authRequired: false
    })

    const serverSpy = {
      route: jest.fn()
    }
    controller.plugin().plugin.register(serverSpy)

    const registeredRoutes = serverSpy.route.mock.calls[0][0]

    expect(registeredRoutes[0].options).toEqual({})
    expect(registeredRoutes[1].options).toEqual({})
  })

  it('should have auth options if auth is required', () => {
    spyOnConfig('featureFlags', {
      authEnabled: true,
      authRequired: true
    })

    const serverSpy = {
      route: jest.fn()
    }
    controller.plugin().plugin.register(serverSpy)

    const registeredRoutes = serverSpy.route.mock.calls[0][0]

    expect(registeredRoutes[0].options).toEqual({
      auth: {
        strategy: 'session',
        mode: 'required'
      }
    })
    expect(registeredRoutes[1].options).toEqual({
      auth: {
        strategy: 'session',
        mode: 'required'
      }
    })
  })

  it('should render an empty application (complete with "cannot start yet" sections)', async () => {
    await session.setState(stateKey, {})

    const { payload, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: controller.urlPath
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    const document = parseDocument(payload)

    expect(statusCode).toBe(statusCodes.ok)
    expect(document.querySelector('#main-content')?.innerHTML).toMatchSnapshot()
  })

  it('should render a partially complete application', async () => {
    await session.setState(stateKey, { origin: validOriginSectionState })

    const { payload, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: controller.urlPath
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    const document = parseDocument(payload)

    expect(statusCode).toBe(statusCodes.ok)
    expect(document.querySelector('#main-content')?.innerHTML).toMatchSnapshot()
  })

  it('should render a complete application', async () => {
    await session.setState(stateKey, {
      origin: validOriginSectionState,
      destination: validDestinationSectionState
    })

    const { payload, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: controller.urlPath
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    const document = parseDocument(payload)

    expect(statusCode).toBe(statusCodes.ok)
    expect(document.querySelector('#main-content')?.innerHTML).toMatchSnapshot()
  })

  it('should respond to a form submission by redirecting the user to the submit path', async () => {
    const { headers, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'POST',
          url: controller.urlPath
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    expect(statusCode).toBe(statusCodes.redirect)
    expect(headers.location).toBe(controller.submitUrlPath)
  })
})
