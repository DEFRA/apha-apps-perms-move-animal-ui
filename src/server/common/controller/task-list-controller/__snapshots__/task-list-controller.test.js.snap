// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskListController should render a complete application 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <h1 class="govuk-heading-l">Test page heading</h1>
          <p data="test-id-task-to-complete-count">
            You have completed all sections.
          </p>
        
<ul class="govuk-task-list">
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/origin/check-answers" aria-describedby="task-list-1-status">
        Movement origin
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-1-status">
      Completed
    </div>
  </li>
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/destination/check-answers" aria-describedby="task-list-2-status">
        Movement destination
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-2-status">
      Completed
    </div>
  </li>
</ul>

        <form method="POST">
          <input type="hidden" name="crumb" value="csrf-value"> <button type="submit" class="govuk-button" data-module="govuk-button">
  Review your application and submit
</button>
        </form>
      </div>
    </div>
  </div>
        "
`;

exports[`TaskListController should render a partially complete application 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <h1 class="govuk-heading-l">Test page heading</h1>
          <p>
            You have
            <span data="test-id-task-to-complete-count">1 out of 2</span>
            sections left to complete.
          </p>
        
<ul class="govuk-task-list">
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/origin/check-answers" aria-describedby="task-list-1-status">
        Movement origin
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-1-status">
      Completed
    </div>
  </li>
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/destination/type-of-destination" aria-describedby="task-list-2-status">
        Movement destination
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-2-status">
      <strong class="govuk-tag govuk-tag--blue">
        Incomplete
      </strong>
    </div>
  </li>
</ul>

        <form method="POST">
          <input type="hidden" name="crumb" value="csrf-value"> <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button">
  Review your application and submit
</button>
        </form>
      </div>
    </div>
  </div>
        "
`;

exports[`TaskListController should render an empty application (complete with "cannot start yet" sections) 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <h1 class="govuk-heading-l">Test page heading</h1>
          <p>
            You have
            <span data="test-id-task-to-complete-count">2 out of 2</span>
            sections left to complete.
          </p>
        
<ul class="govuk-task-list">
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/origin/to-or-from-own-premises" aria-describedby="task-list-1-status">
        Movement origin
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-1-status">
      <strong class="govuk-tag govuk-tag--blue">
        Incomplete
      </strong>
    </div>
  </li>
  <li class="govuk-task-list__item">
    <div class="govuk-task-list__name-and-hint">
      <div>
        Movement destination
      </div>
    </div>
    <div class="govuk-task-list__status govuk-task-list__status--cannot-start-yet" id="task-list-2-status">
      Cannot start yet
    </div>
  </li>
</ul>

        <form method="POST">
          <input type="hidden" name="crumb" value="csrf-value"> <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button">
  Review your application and submit
</button>
        </form>
      </div>
    </div>
  </div>
        "
`;
